samples/hssi/hssi.cpp

```diff
@@ -0,0 +1,654 @@
+//   without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
+// IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+#include <iostream>
+#include <string>
+#include <sstream>
+#include <exception>
+#include <csignal>
+#include <cstring>
+#include <cstdio>
+#include <unistd.h>
+#include <regex.h>
+#include <glob.h>
+#include <CLI/CLI.hpp>
+#include <opae/cxx/core/handle.h>
+#include <opae/cxx/core/token.h>
+#include <opae/cxx/core/properties.h>
+#include <opae/cxx/core/sysobject.h>
+#include "hssi_mbox.h"
+
+using namespace opae::fpga::types;
+
+#define AFU_ID "823c334c-98bf-11ea-bb37-0242ac130002"
+
+#define DEFAULT_ETH_LOOPBACK std::string("on")
+#define DEFAULT_PACKETS 1
+#define DEFAULT_PACKET_LENGTH 64
+#define DEFAULT_QSFP_PORT 0
+#define DEFAULT_RANDOM_LENGTH std::string("fixed")
+#define DEFAULT_RANDOM_PAYLOAD std::string("incremental")
+#define DEFAULT_RND_SEED0 0x5eed0000
+#define DEFAULT_RND_SEED1 0x5eed0001
+#define DEFAULT_RND_SEED2 0x00025eed
+
+#define ETH_AFU_DFH           0x0000
+#define ETH_AFU_ID_L          0x0008
+#define ETH_AFU_ID_H          0x0010
+#define TRAFFIC_CTRL_CMD      0x0030
+#define TRAFFIC_CTRL_DATA     0x0038
+#define TRAFFIC_CTRL_PORT_SEL 0x0040
+#define AFU_SCRATCHPAD        0x0048
+
+#define CSR_NUM_PACKETS     0x3c00
+#define CSR_RANDOM_LENGTH   0x3c01
+#define CSR_RANDOM_PAYLOAD  0x3c02
+#define CSR_START           0x3c03
+#define CSR_STOP            0x3c04
+#define CSR_SRC_ADDR0       0x3c05
+#define CSR_SRC_ADDR1       0x3c06
+#define CSR_DEST_ADDR0      0x3c07
+#define CSR_DEST_ADDR1      0x3c08
+#define CSR_PACKET_TX_COUNT 0x3c09
+#define CSR_RND_SEED0       0x3c0a
+#define CSR_RND_SEED1       0x3c0b
+#define CSR_RND_SEED2       0x3c0c
+#define CSR_PACKET_LENGTH   0x3c0d
+
+#define CSR_NUM_PKT         0x3d00
+#define CSR_PKT_GOOD        0x3d01
+#define CSR_PKT_BAD         0x3d02
+#define CSR_AVST_RX_ERR     0x3d07
+
+#define CSR_MAC_LOOP        0x3e00
+
+#define MBOX_TIMEOUT 1000
+
+bool parse_filter_address(properties::ptr_t filter, const char *sbdf)
+{
+  regex_t re;
+  regmatch_t matches[5];
+
+  memset(&matches, 0, sizeof(matches));
+
+  std::string sbdf_expr("([0-9a-fA-F]{4}):([0-9a-fA-F]{2}):([0-9a-fA-F]{2})\\.([0-7])");
```

Comment:
You can simplify this by using this pattern:
```c++
 std::string sbdf_expr("(([0-9a-fA-F]{4}):)?([0-9a-fA-F]{2}):([0-9a-fA-F]{2})\\.([0-7])");
```

That gives you two more groups but they'll start at index 2.
```c++
if (matches[2].rm_so > -1) {
  std::string segment(sbdf + matches[2].rm_so, matches[2].rm_eo - matches[2].rm_so);
  filter->segment = static_cast<uint16_t>(std::stoul(segment, nullptr, 16));
} else {
  filter->segment = 0;
}

```

Reply:
As is it reads a little more cleanly, I think.